src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilter.java

```diff
@@ -0,0 +1,334 @@
+import scala.Tuple2;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.util.*;
+
+/**
+ * Performs PathSeq filtering steps and manages associated resources.
+ */
+public final class PSFilter implements AutoCloseable {
+
+    private final JavaSparkContext ctx;
+    private final PSFilterArgumentCollection filterArgs;
+    private final MetricsState metricsState;
+    private JavaRDD<GATKRead> reads;
+    private final SAMFileHeader header;
+
+    public PSFilter(final JavaSparkContext ctx, final PSFilterArgumentCollection filterArgs, final JavaRDD<GATKRead> inputReads,
+                    final SAMFileHeader header) {
+        Utils.nonNull(ctx);
+        Utils.nonNull(filterArgs);
+        Utils.nonNull(inputReads);
+        this.ctx = ctx;
+        this.filterArgs = filterArgs;
+        this.metricsState = initializeMetics(filterArgs.metricsFileUri);
+        this.reads = inputReads;
+        this.header = header;
+    }
+
+    @VisibleForTesting
+    static JavaRDD<GATKRead> setPairFlags(final JavaRDD<GATKRead> reads, final int readsPerPartitionGuess) {
+        return repartitionPairedReads(reads).mapPartitions(iter -> setPartitionUnpairedFlags(iter, readsPerPartitionGuess));
+    }
+
+    private static JavaRDD<GATKRead> clearAllAlignments(final JavaRDD<GATKRead> reads, final SAMFileHeader header) {
+        return reads.map(read -> clearReadAlignment(read, header));
+    }
+
+    /**
+     * Returns input read with alignment-related info cleared
+     */
+    private static GATKRead clearReadAlignment(final GATKRead read, final SAMFileHeader header) {
+        final GATKRead newRead = new SAMRecordToGATKReadAdapter(new SAMRecord(header));
+        newRead.setName(read.getName());
+        newRead.setBases(read.getBases());
+        newRead.setBaseQualities(read.getBaseQualities());
+        newRead.setIsUnmapped();
+        newRead.setIsPaired(read.isPaired());
+        if (read.isFirstOfPair()) {
+            newRead.setIsFirstOfPair();
+        } else if (read.isSecondOfPair()) {
+            newRead.setIsSecondOfPair();
+        }
+        final String readGroup = read.getReadGroup();
+        if (readGroup != null) {
+            newRead.setAttribute(SAMTag.RG.name(), readGroup);
+        }
+        return newRead;
+    }
+
+    /**
+     * Sets proper pairedness flags
+     */
+    private static Iterator<GATKRead> setPartitionUnpairedFlags(final Iterator<GATKRead> iter, final int readsPerPartitionGuess) {
+        final Tuple2<List<GATKRead>, List<GATKRead>> lists = getPairedAndUnpairedLists(iter, readsPerPartitionGuess);
+        final List<GATKRead> pairedReadsList = lists._1;
+        final List<GATKRead> unpairedReadsList = lists._2;
+        for (final GATKRead unpairedRead : unpairedReadsList) {
+            unpairedRead.setIsPaired(false);
```

Comment:
What I decided on was to run all the reads back through clearReadAlignment() to clean that stuff up. It's a little hacky but at this point we're down to a relatively small number of reads where I'm not concerned about performance. (I thought it was written this way but I must have done this on another branch somewhere and forgotten, I've added it now).

Also realized that I should only set firstInPair/secondInPair in clearReadAlignment() if the read is paired so I've added a check for that too.